src/metal_guardian.c

```diff
@@ -0,0 +1,476 @@
+    sub_080708DC(&laser->base, &laser->base.unk10, 0xc, 0x311, 7, 0x1b);
+    laser->base.unk10.unk1F = 0;
+    if (gKirbys[gUnk_0203AD3C].base.base.base.roomId == laser->base.roomId) {
+        laser->base.unk10.unk1F = sub_0803DF24(0x311);
+        if (laser->base.unk10.unk1F == 0xff) {
+            laser->base.unk10.unk1F = sub_0803DFAC(0x311, 0);
+        }
+    }
+    else {
+        laser->base.unk10.unk1F = 0;
+    }
+    PlaySfx(&laser->base, 0x139);
+}
+
+static u32 sub_080BE0E8(struct MetalGuardianLaser* arg0) {
+    struct Sprite sprite;
+    arg0->base.flags |= 4;
+    if (gKirbys[gUnk_0203AD3C].base.base.base.roomId == arg0->base.roomId) {
+        if (arg0->base.flags & 0x4000) {
+            if (arg0->base.unk10.unk0 == 0) {
+                struct Sprite *sprite2;
+                (sprite2 = &arg0->base.unk10)->unk0 = sub_0803DE54(0xc, arg0->base.unk10.unkC, arg0->base.unk10.unk1A);
+                sprite2->unk8 = arg0->base.unk10.unk8 & ~0x80000;
+                CpuCopy32(sprite2, &sprite, sizeof(struct Sprite));
+                sub_0815521C(&sprite, arg0->base.unk1);
+                sprite2->unk8 = arg0->base.unk10.unk8 | 0x80000;
+            }
+        }
+        else {
+            if (arg0->base.unk10.unk0 == 0) {
+                struct Sprite *sprite2;
+                (sprite2 = &arg0->base.unk10)->unk0 = sub_081570B0(0xc);
+                sprite2->unk8 = arg0->base.unk10.unk8 & ~0x80000;
+                CpuCopy32(&arg0->base.unk10, &sprite, sizeof(struct Sprite));
+                sub_0815521C(&sprite, arg0->base.unk1);
+            }
+        }
+    }
+    else {
+        if (arg0->base.unk10.unk0 != 0) {
+            if (!(arg0->base.flags & 0x4000)) {
+                sub_08157190(arg0->base.unk10.unk0);
+                arg0->base.unk10.unk0 = 0;
+            }
+        }
+        arg0->base.unk10.unk8 |= 0x80000;
+    }
+    if (gKirbys[gUnk_0203AD3C].base.base.base.roomId == arg0->base.roomId) {
+        if (arg0->base.unk10.unk1F == 0) {
+            arg0->base.unk10.unk1F = sub_0803DF24(0x311);
+            if (arg0->base.unk10.unk1F == 0xff)  {
+                arg0->base.unk10.unk1F = sub_0803DFAC(0x311, 0);
+            }
+        }
+    }
+    else {
+        arg0->base.unk10.unk1F = 0;
+    }
+    return 0;
+}
+
+static void sub_080BE228(struct Object2* arg0) {
+    ObjectSetFunc(&arg0->base, 2, sub_080BD634);
+    arg0->base.xspeed = 0;
+    arg0->base.yspeed = 0;
+    arg0->base.flags |= 0x40;
+    arg0->base.flags &= ~0x20;
+    arg0->base.counter = 0x5a;
+}
+
+static void sub_080BE25C(struct Object2* arg0) {
+    ObjectSetFunc(&arg0->base, 0, sub_080BD988);
+    arg0->base.xspeed = 0;
+    arg0->base.flags &= ~0x40;
+}
+
+static void sub_080BE284(struct Object2* arg0) {
+    ObjectSetFunc(&arg0->base, 1, sub_080BDA00);
+    arg0->base.flags |= 0x40;
+    arg0->base.flags &= ~2;
+}
+
+static u32 sub_080BE2A8(struct MetalGuardianLaser* arg0) {
+    sub_08073D2C((struct Object2*)arg0);
```

Comment:
Change the function prototype into `void sub_08073D2C(struct ObjectBase *);` instead